pkg/maintenance/alertmanager.go

```diff
@@ -0,0 +1,173 @@
+package maintenance
+
+import (
+	"context"
+	"github.com/go-openapi/runtime"
+	httptransport "github.com/go-openapi/runtime/client"
+	"github.com/go-openapi/strfmt"
+	routev1 "github.com/openshift/api/route/v1"
+	amSilence "github.com/prometheus/alertmanager/api/v2/client/silence"
+	amv2Models "github.com/prometheus/alertmanager/api/v2/models"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"net/http"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+	"time"
+)
+
+var (
+	alertManagerNamespace          = "openshift-monitoring"
+	alertManagerRouteName          = "alertmanager-main"
+	alertManagerServiceAccountName = "prometheus-k8s"
+	alertManagerBasePath           = "/api/v2/"
+)
+
+type alertManagerMaintenance struct {
+	client    client.Client
+	transport *httptransport.Runtime
+}
+
+func newAlertManagerMaintenance(client client.Client) (Maintenance, error) {
+	transport, err := getTransport(client)
+	if err != nil {
+		return nil, err
+	}
+
+	transport.DefaultAuthentication, err = getAuthentication(client)
+	if err != nil {
+		return nil, err
+	}
+
+	return &alertManagerMaintenance{
+		client:    client,
+		transport: transport,
+	}, nil
+}
+
+func getTransport(c client.Client) (*httptransport.Runtime, error) {
+	amRoute := &routev1.Route{}
+	err := c.Get(
+		context.TODO(),
+		types.NamespacedName{Namespace: alertManagerNamespace, Name: alertManagerRouteName},
+		amRoute,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return httptransport.New(
+		amRoute.Spec.Host,
+		alertManagerBasePath,
+		[]string{"https"},
+	), nil
+}
+
+func getAuthentication(c client.Client) (runtime.ClientAuthInfoWriter, error) {
+	sl := &corev1.SecretList{}
+	err := c.List(
+		context.TODO(),
+		sl,
+		&client.ListOptions{Namespace: alertManagerNamespace},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	var token string
+	for _, s := range sl.Items {
+		if strings.Contains(s.Name, alertManagerServiceAccountName+"-token") {
+			token = string(s.Data["token"])
+		}
+	}
+
+	return httptransport.BearerToken(token), nil
+}
+
+func (amm *alertManagerMaintenance) Start() error {
+	createdBy := "managed-upgrade-operator"
+	comment := "Test comment"
+	startsAt := strfmt.DateTime(time.Now().UTC())
+	endsAt := strfmt.DateTime(time.Now().Add(time.Minute * 5))
+
+	name := "alertname"
+	isRegex := true
+	value := "[A-Z].*"
```

Comment:
We need make sure this only silence the alerts not upgrade related

Reply:
I agree, was going to make the same comment. Can we try and make it exclude the existing "ClusterUpgradingSRE" alert from the silence? And maybe add more later if we need to.

Reply:
Yeah i'll refine these when I rebase on haorans work :+1: